contracts/Lockbox.cdc

```diff
@@ -0,0 +1,329 @@
+            self.nodeStaker = nil
+            self.nodeDelegator = nil
+            self.unlockLimit = 0.0
+        }
+
+        // FungibleToken.Receiver actions
+
+        pub fun deposit(from: @FungibleToken.Vault) {
+            let balance = from.balance
+
+            self.vault.deposit(from: <- from)
+
+            self.increaseUnlockLimit(delta: balance)
+        }
+
+        // FungibleToken.Provider actions
+
+        pub fun withdraw(amount: UFix64): @FungibleToken.Vault {
+            pre {
+                self.unlockLimit >= amount: "Requested amount exceeds unlocked token limit"
+            }
+
+            post {
+                self.unlockLimit == before(self.unlockLimit) - amount: "Updated unlocked token limit is incorrect"
+            }
+
+            let vault <- self.vault.withdraw(amount: amount)
+
+            self.decreaseUnlockLimit(delta: amount)
+
+            return <- vault
+        }
+
+        access(self) fun decreaseUnlockLimit(delta: UFix64) {  
+            self.unlockLimit = self.unlockLimit - delta
+        }
+
+        // Lockbox.TokenAdmin actions
+
+        pub fun increaseUnlockLimit(delta: UFix64) {  
+            self.unlockLimit = self.unlockLimit + delta
+        }
+
+        // Lockbox.TokenHolder actions
+
+        pub fun registerNode(id: String, role: UInt8, networkingAddr: String, networkingKey: String, stakingKey: String) {
+            self.nodeStaker <- FlowIDTableStaking.addNodeRecord(self.nodeInfo)
+        }
+
+        pub fun registerDelegator(nodeID: String) {
+            self.nodeDelegator <- FlowIDTableStaking.registerDelegator(nodeID: nodeID)
+        }
+
+        pub fun createTokenHolder(tokenManager: Capability<&LockedTokenManager>): @TokenHolder {
+            return <- create TokenHolder(tokenManager: tokenManager)
+        }
+    }
+
+    // stored in Holder unlocked account
+    pub resource TokenHolder {
+
+        access(self) var tokenManager: Capability<&LockedTokenManager>
+
+        access(self) var nodeStakerProxy: LockedNodeStakerProxy?
+
+        access(self) var nodeDelegatorProxy: LockedNodeDelegatorProxy?
+
+        init(tokenManager: Capability<&LockedTokenManager>) {
+            pre {
+                tokenManager.borrow() != nil: "Invalid capability"
+            }
+            self.tokenManager = tokenManager
+            self.nodeStakerProxy = nil
+            self.nodeDelegatorProxy = nil
+        }
+
+        pub fun deposit(from: @FungibleToken.Vault) {
+            let tokenManagerRef = self.tokenManager.borrow()!
+
+            tokenManagerRef.deposit(from: <-from)
+        }
+
+        // FungibleToken.Provider actions
+
+        pub fun withdraw(amount: UFix64): @FungibleToken.Vault {
+            let tokenManagerRef = self.tokenManager.borrow()!
+
+            return <- tokenManagerRef.withdraw(amount: amount)
+        }
+
+        pub fun createNodeStaker(id: String, role: UInt8, networkingAddr: String, networkingKey: String, stakingKey: String) {
```

Comment:
Hmm, should `role` be part of `StakingProxy.NodeInfo`? And then maybe these params can just be replaced with a single `nodeInfo` param?

```suggestion
        pub fun createNodeStaker(id: String, role: UInt8, networkingAddr: String, networkingKey: String, stakingKey: String) {
```

Reply:
Yeah you're right. I'll replace those